### ISSUE TITLE
Dataset::write taking trait object makes it difficult to work with in an async context

### ISSUE THREAD

**Issue by dnsco** [lancedb/lance#1031]
I have a method on an ingestor struct:

``` rust
pub async fn write(&self, buffer: TemporalBuffer) -> Result<Dataset> {
        let mut reader: Box<dyn RecordBatchReader> =
            Box::new(RecordBatchBuffer::new(buffer.batches));

        let dataset =
            Dataset::write(&mut reader, self.filename.as_ref(), Some(self.write_params)).await?;

        Ok(dataset)
    }
```

upstream when I have a task that listens for these buffers on a channel and passes them to the ingestor in an infinite loop:

```rust
let mut tasks = Joinset::new()
// ...
tasks.spawn(async move {
        loop {
            let buf = rx_buffer.recv().await?;
            ingestor.write(buf).await?;
        }
});
```

Because Dataset::write takes a trait object that doesn't require a Send bound I have a nightmare because Joinset::spawn requires its futures are ```Send```.

I'll confess I'm not the best at this level of rust, but it seems like the issue would disappear if write used static dispatch instead of dynamic dispatch ```Dataset::write<R: RecordBatchReader>(...)```

If anyone else has any ideas for how to mitigate this I'm open to suggestions.

Thanks!

**Comment by wjones127** [comment]
Seems reasonable. I think what we want to do is make the future that `Dataset::write()` returns `impl Future<Output = ...> + Send`. This means we'll add `Send` requirement to the `reader` arg too.


I think I'd also like to change it so it consumes the record batch reader instead of take a mutable reference to it, so it doesn't have to be kept alive across an await point.

**Comment by wjones127** [comment]
The only reason we don't have the `Send` bound on the `reader` argument is that arrow-rs didn't mark the FFI Arrow streams as `Send`. That's fixed in the next version of Arrow, but we need to wait until the next DataFusion release to upgrade Arrow. So blocked on that.

The next DataFusion release is being voted on now, so it should only be a couple days until we can upgrade.

**Comment by dnsco** [comment]
Do you have any ideas for how we can work around this until then?

**Comment by dnsco** [comment]
My understanding which could very well be mistaken is that the only reason the Send bound is required is because we're using a trait object and not monomorphicizing the function because trait objects are !Send unless otherwise specified so if the function didn't take a trait object and took a B: RecordBatchReader instead the future would be send as long as the B were send.

&mut Box<dyn ...> is a couple levels of indirection that I don't understand the utility of, the problem could be solved by adding + Send  to the mix, but is there a reason for it to exist in this way in the first place?

Unfortunately, there's not an implementation of RecordBatchReader for Box<B> where B: RecordBatchReader in arrow so it made it hard for me to experiement but...

tl;dr why the trait object? as a user of this library they force this kind of papercut and I just lost a few hours being sad :'(

**Comment by wjones127** [comment]
> tl;dr why the trait object?

We could make it generic I suppose, but I suspect it will likely increase compile times due to the monomorphization. Each instance will generate a lot of code, without much performance benefit.

Regardless, I think your issue is simply that `Dataset::write()` returns a Future that isn't `Send`. Even if we changed to be generic, we'd still have to add a `Send` bound to it (I don't think the compiler would automatically propagate that for specific monomorphizations).

I know this is a painful API, and it's been something I've been wanting to change for a while. Hence why I made the upstream change to make this possible for us. We just need a few more days to get that in.

> Do you have any ideas for how we can work around this until then?

Would using `spawn_local` ([docs](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.spawn_local)) be acceptable for the time being?

**Comment by dnsco** [comment]
I'll take a crack at it with spawn_local next week. My pair has logged off with the changes on their computer for the weekend. Thanks for finding that.

And thanks for working on the changes upstream. I was kinda erring on the assumption it wouldn't add too much to compile time on the assumption that most projects wouldn't call it with multiple types, but I guess it could bloat lance's test times or something. I don't suspect it will have much performance benefit, but do suspect a strong ergonomic benefit, but am most likely mistaken.

I was operating on the assumption that the only reason the future is unsend is because it holds a trait object, and trait objects are unsend, but if it's unsend for other reason, I'll thank you again for doing all this work.

Thanks for your help, have a great weekend!
